include/zephyr/bluetooth/gatt.h

```diff
@@ -167,16 +167,57 @@ typedef ssize_t (*bt_gatt_attr_write_func_t)(struct bt_conn *conn,
 
 /** @brief GATT Attribute structure. */
 struct bt_gatt_attr {
-	/** Attribute UUID */
+	/** @brief Attribute Type, aka. "UUID" */
 	const struct bt_uuid *uuid;
-	/** Attribute read callback */
+
+	/** @brief Attribute Value function
+	 *
+	 * This is the general API to read from the Attribute Value of
+	 * this Attribute in on-air format.
+	 *
+	 * This function may safely assume the Attribute Permissions are
+	 * satisfied for this read.
+	 *
+	 * This function is primarily invoked by the GATT server when a
+	 * remote GATT client performs a read, but it can also be
+	 * invoked for a local read.
+	 *
+	 * May be NULL if the attribute is not readable.
+	 */
 	bt_gatt_attr_read_func_t read;
-	/** Attribute write callback */
+
+	/** @brief Attribute Value write method
+	 *
+	 * This is the general API to write to the Attribute Value of
+	 * this Attribute in on-air format.
+	 *
+	 * This function may safely assume the Attribute Permissions are
+	 * satisfied for this write.
+	 *
+	 * May be NULL if the attribute is not writable.
+	 */
 	bt_gatt_attr_write_func_t write;
-	/** Attribute user data */
+
+	/** @brief Userdata for the read() and write()
+	 *
+	 * The meaning of this field is not specified here.
```

Comment:
Maybe something like
```suggestion
	 * The meaning of this field is context and state specific.
	 * When the attribute is a local attribute, this parameter contains
	 * user provided data.
	 * When the attribute is a remote attribute obtained through service discovery,
	 * this parameter contains temporary information about the discovered attribute.
	 * See @ref bt_gatt_discover_func_t for more details.
```
?
Otherwise users will probably ask themselves where it is specified

Reply:
I think we should avoid the terms 'local attribute' and 'remote attribute' here. These terms would need to be defined, and for the purpose of defining user_data, the struct itself is just a data object with some methods to read and write.

I don't see how "context and state specific" conveys any more meaning than "not specified here". I think it look even more mysterious.

I did several iterations of how to explain the meaning of this field. At the end it just ended up explaining common sense contract-based programming, and it made no sense to put that explanation here.

> Otherwise users will probably ask themselves where it is specified

That would be the API you get the pointer from. It determines the contract.

What if it says nothing?

Use the C types and common sense. This field is a void pointer, so it's opaque and not meant for you. Why don't you want to use the generic read method?

Is that it?

Well, no. You can allow yourself to infer contracts. If e.g. a read method implementation publishes a contract, and that contract gives some public requirements about the layout of the user_data, and this attribute object uses this read method, then you can infer that this contract is valid whenever the read method is invoked.

And a common sense default is that if a method is not NULL and you have no other contract saying otherwise, then you may call this method. And, since you might call this method, this implies the contract holds at all times you see that method. E.g. i.e. The read method function pointer can be used as a marker to determine the layout of the user data.

Here you can argue that we should also make sure the read-bit is set in the attribute permissions before we assume the read method may be invoked. This also assumes the object is fully initialized. But I think common sense says the read method is always a marker. Same is true for the write method.

There is nothing specific to GATT attributes here, so again, I'm not sure we should be explaining this here.

I'm not against adding some examples of how the user data may be used. Would that work?

---

> \+ When the attribute is a local attribute, this parameter contains user provided data.

I read this as "Normally user data is user data." It's the discovery API that's weird for using the user data field as a place to store results. Maybe we could add some safety-functions that cast the data after a check somehow. Or redo the whole API.

Reply:
I see your point. I do not have a better suggestion at the moment